Bio/SeqUtils/__init__.py

```diff
@@ -667,6 +668,61 @@ def calculate(self, sequence):
 
         return exp(cai_value / cai_length)
 
+    def optimize(self, sequence, seq_type="DNA", strict=False):
+        """Return a new DNA sequence with preferred codons only.
+
+        Uses the codon adaptiveness table defined by the CodonAdaptationIndex
+        object to generate DNA sequences with only preferred codons.
+        May be useful when designing DNA sequences for transgenic protein
+        expression or codon-optimized proteins like fluorophores.
+
+        Arguments:
+            - sequence: DNA, RNA, or protein sequence to codon-optimize.
+                        Supplied as a str, Seq, or SeqRecord object.
+            - seq_type: String specifying type of sequence provided.
+                        Options are "DNA", "RNA", and "protein". Default is "DNA".
+            - strict:   Determines whether an exception should be raised when
+                        two codons are equally prefered for a given amino acid.
+        Returns:
+            Seq object with DNA encoding the same protein as the sequence argument,
+            but using only preferred codons as defined by the codon adaptation index.
+            If multiple codons are equally preferred, a warning is issued
+            and one codon is chosen for use in the optimzed sequence.
+        """
+        try:  # If seq record is provided, convert to sequence
+            sequence = sequence.seq
+        except AttributeError:  # not a  SeqRecord object
+            pass
+        seq = sequence.upper()
+        # Make dict with amino acids referencing preferred codons
+        pref_codons = {}
+        for codon, aminoacid in self._table.forward_table.items():
+            if self[codon] == 1.0:
+                if aminoacid in pref_codons:
+                    msg = f"{pref_codons[aminoacid]} and {codon} are equally preferred."
+                    if strict:
+                        raise Exception(msg)
+                    warnings.warn(f"{msg} Using {codon}", RuntimeWarning)
```

Comment:
Do you still need this warning?

Reply:
The use case I imagine is evaluating codon preferences from an organisms coding sequences or ribosomal genes, then using those codon preferences to codon-optimize a DNA sequence or set of sequences for expression in the organism of interest. 

Equally preferred codons would only be likely when using a single, or a small number of sequences to generate the CAI table. If I were using this I would always want to be made aware of the fact that there are equally preferred codons and that one was used over the other in generating the optimized sequence. I can imagine that someone might want to throw an exception instead, but I do like keeping the warning as default behavior.

Reply:
The problem with warnings is that they are unreliable. By default, they get triggered only once, and many users won't be aware of this. It gives a false sense of security.

> If I were using this I would always want to be made aware of the fact that there are equally preferred codons and that one was used over the other in generating the optimized sequence.

You can do this by setting `strict=True`, and if the exception is raised, you can decide whether you want to repeat the calculation with `strict=False`. Effectively it's the same as using a warning, but this way it is reliable and reproducible.

Reply:
So @mdehoon would you favour strict=True by default with an exception, and strict=False with no warnings?